tuxemon/npc.py

```diff
@@ -847,5 +847,87 @@ def alter_item_quantity(
 
         return success
 
+    def can_buy_item(self, item_slug: str, qty: int, unit_price: int) -> bool:
+        current_money = self.game_variables.get("money")
+        if current_money is not None:
+            return current_money >= qty * unit_price
+        # If no 'money' variable, must be an NPC. Always allow buying:
+        return True
+
+    def can_sell_item(self, item_slug: str, qty: int, unit_price: int) -> bool:
+        current_item = self.inventory.get(item_slug)
+        if current_item:
+            current_qty = current_item["quantity"]
+            return current_qty >= qty or current_item["infinite"]
+        # If they don't have the item, then they can't sell the item:
+        return False
+
+    def buy_decrease_money(
+        self,
+        session: Session,
+        seller: NPC,
+        item_slug: str,
+        qty: int,
+        unit_price: int,
+    ) -> None:
+        """Decreases current money during a buy transaction, but doesn't change
+        an item's quantity (use NPC.give_item too after this)
+
+        Raises an exception if there's not enough money to pay the price."""
+
+        # Only players will have money in game_variables, so only update money or raise
+        # exception if it exists:
+        if self.game_variables.get("money"):
+            if not self.can_buy_item(item_slug, qty, unit_price):
+                raise Exception(
+                    f"Tried to buy item with {self.game_variables['money']} coins "
+                    f"but not enough money:\n"
+                    f"price {unit_price} times qty {qty} is {qty * unit_price}"
+                )
+
+            self.game_variables["money"] -= qty * unit_price
+
+    def sell_increase_money(
+        self,
+        session: Session,
+        buyer: NPC,
+        item_slug: str,
+        qty: int,
+        unit_price: int,
+    ) -> None:
+        """Increases current money during a sell transaction, but doesn't change
+        an item's quantity (use NPC.give_item too after this)
+
+        Raises an exception if there's not enough items in the inventory."""
+        current_item = self.inventory.get(item_slug)
+        if not current_item or not self.can_sell_item(
+            item_slug, qty, unit_price
+        ):
+            raise Exception(
+                f"Tried to sell item of qty {qty}, but not enough available."
+            )
+
+        # Only players will have money in game_variables, so only update money
+        # if it exists:
+        if self.game_variables.get("money"):
+            self.game_variables["money"] += qty * unit_price
```

Comment:
What if we sell an item whilst having no money?

Reply:
> What if we sell an item whilst having no money? 

Good catch - I've fixed this now (I'll admit I haven't tested the player selling anything yet - this PR only adds buying for the player. This will help fix the bugs in the selling when I do that too, though!)